### DISCUSSION TITLE
How to apply a mask with transparent background?

### DISCUSSION THREAD

**Discussion by TheColorRed** [dlemstra/Magick.NET#876] [Help] [answered]
I am trying to take an image and fade the edges. What I have done is created an elliptical gradient (black in the center and white on the edges). I have tried many things nothing seems to be working. I was able to get this once but it was on the `image` layer and not the one that I wanted it on causing issues.

The below creates a 100% transparent image and places it on the final image so basically the image disappears. I have tried using `White` for the `maskedImage`, but that leaves an image with a white edge and not a transparent one.

```csharp
var image = new MagickImage("images/background.png");
var boat = new MagickImage("images/boat.png");
var mask = new MagickImage("images/mask.png");

// Create the masked image
var maskedImage = new MagickImage(MagickColors.Transparent, boat.Width, boat.Height);
maskedImage.SetWriteMask(mask);
maskedImage.Composite(boat, Channels.RGB | Channels.Alpha);

// I have tried removing the mask at this point

// Add the masked image to the final image
image.Composite(maskedImage, Gravity.Center, CompositeOperator.Overlay, Channels.RGB);

// I have tried removing the mask at this point

image.Write("out.png");
```
**background.png**
![background](https://user-images.githubusercontent.com/1571806/110806549-c9829800-8247-11eb-9fd4-8599d935a9db.png)

**boat.png**
![boat](https://user-images.githubusercontent.com/1571806/110807196-68a78f80-8248-11eb-991f-4274f7df9a09.png)

**mask.png**
![mask](https://user-images.githubusercontent.com/1571806/110807686-e9ff2200-8248-11eb-897e-a6d5a4beba11.png)

**Comment by dlemstra** [comment, marked as answer]
Have you tried removing the write mask from `layerImage`? And is it possible to make your code sample smaller and include some test images?

**Reply by TheColorRed** [reply]
@dlemstra I have made updates to the OP.

**Reply by TheColorRed** [reply]
And yes, I have tried removing the mask from the image.

**Reply by TheColorRed** [reply]
If I put the mask on `image`, and write `boat` to it, then it works. However, I cannot set gravity on the boat because it won't be behind the mask.

```csharp
var image = new MagickImage("images/background.png");
var boat = new MagickImage("images/boat.png");
var mask = new MagickImage("images/mask.png");

image.SetWriteMask(mask);

image.Composite(boat, CompositeOperator.Overlay, Channels.RGB | Channels.Alpha);

image.Write("out.png");
```

**Reply by dlemstra** [reply]
Is this what you are looking for:

```C#
mask.Extent(image.Width, image.Height, Gravity.Center);

image.SetWriteMask(mask);

image.Composite(boat, Gravity.Center);
```

**Reply by TheColorRed** [reply]
That looks like it does what I am looking for! Thanks!

**Reply by TheColorRed** [reply]
Is there anyway for this to work with a `Transparent` root image? This gives me a blank image...

```csharp
var image = new MagickImage(MagickColors.Transparent, 800, 600);
var boat = new MagickImage("images/boat.png");
var mask = new MagickImage("images/mask.png");

mask.Extent(image.Width, image.Height, Gravity.Center);

image.SetWriteMask(mask);

image.Composite(boat, Gravity.Center, Channels.RGB | Channels.Alpha);

image.Write("out.png");
```

**Reply by dlemstra** [reply]
You will need a slightly different approach for that:

```C#
image.Composite(boat, Gravity.Center, CompositeOperator.Over);

mask.Extent(image.Width, image.Height, Gravity.Center);
mask.Negate();

image.Composite(mask, CompositeOperator.CopyAlpha);
```

**Reply by TheColorRed** [reply]
After a few hours of trial and error of it working in my testing environment and not in my actual application, I noticed that it wasn't working in my application when using the following to generate a gradient mask.

```csharp
var start = MagickColors.Black.ToString();
var end = MagickColors.White.ToString();
var readSettings = new MagickReadSettings {
  Width = boat.Width,
  Height = boat.Height
};

readSettings.SetDefine("gradient:extent", "ellipse");
var mask = new MagickImage($"radial-gradient:{start}-{end}", readSettings);
```

When `Negate()` is run, the mask is converted to complete transparency along with the boat. When I don't `Negate()` the boat is 100% opaque.

I have tried many different settings on the gradient, but none of them work. Visually the gradient is created properly.

**Note:** The radial image I supplied in the OP was generated with the gimp, and that image works.

**Reply by TheColorRed** [reply]
I got it, just needed to turn off the alpha. Thanks for the help!

```csharp
mask.Alpha(AlphaOption.Off);
mask.Negate();
```
